signal.go

```diff
@@ -38,17 +45,93 @@ func (sig ShutdownSignal) String() string {
 }
 
 func newSignalReceivers() signalReceivers {
-	return signalReceivers{notify: signal.Notify}
+	return signalReceivers{
+		notify:  signal.Notify,
+		signals: make(chan os.Signal, 1),
+	}
 }
 
 type signalReceivers struct {
-	m      sync.Mutex
-	last   *ShutdownSignal
-	done   []chan os.Signal
-	notify func(c chan<- os.Signal, sig ...os.Signal)
+	m        sync.Mutex
+	signals  chan os.Signal
+	shutdown chan struct{}
+	finished chan struct{}
+	notify   func(c chan<- os.Signal, sig ...os.Signal)
+
+	last *ShutdownSignal
+	done []chan os.Signal
+	wait []chan ShutdownSignal
+}
+
+func (recv *signalReceivers) relayer(ctx context.Context) {
+	defer func() {
+		select {
+		case recv.finished <- struct{}{}:
+		default:
+		}
```

Comment:
I believe this whole thing can just be a simple send?
```suggestion
	recv.finished <- struct{}{}
```